google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -154,32 +163,91 @@ def reset
 
         def close
           @thread.kill if @thread
-          @all_sessions.each(&:delete_session)
-          @all_sessions = []
-          @session_queue = []
-          @transaction_queue = []
+          # Unblock all waiting threads
+          @closed = true
+          @resource.broadcast
+          # Delete all sessions
+          @mutex.synchronize do
+            @all_sessions.map do |s|
```

Comment:
There is a chance of session leak here. What if there are multiple threads creating sessions at this point. Those sessions would not yet be in all_sessions so would never get deleted.